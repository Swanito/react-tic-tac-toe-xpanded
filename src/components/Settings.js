import React, { useContext, useState } from 'react';
import { GameContext } from '../context/GameContext';

const styles = {
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'flex-start',
  width: '20em',
  margin: '0 auto',
};

export default function Settings() {
  const { boardSize, setBoardSize } = useContext(GameContext);

  const [selectedSize, setSelectedSize] = useState(boardSize);

  const availableBoardSizes = [
    { label: '3*3', size: 3 },
    { label: '6*6', size: 6 },
    { label: '9*9', size: 9 },
  ];

  const handleSubmit = (e) => {
    setBoardSize(Number(selectedSize));
  };

  const handleSizeChange = (e) => {
    setSelectedSize(e.target.value);
  }

  return (
    <GameContext.Consumer>
      {({ boardSize, players, currentTurnPlayer, winner }) => {
        return (
          <div style={styles}>
            <div>
              PLAYERS: {players} - BOARD SIZE: {boardSize}*{boardSize}
            </div>
            <br />
            <div>IT'S YOUR TURN {currentTurnPlayer}</div>
            <br />
            <div>
              <form onSubmit={handleSubmit}>
                <label htmlFor="board-size-select">Change board Size</label>
                <select value={selectedSize} onChange={handleSizeChange}>
                  {availableBoardSizes.map((option) => (
                    <option value={option.size} key={option.size}>
                      {option.label}
                    </option>
                  ))}
                </select>
                <button type="submit">{winner ? 'Restart' : 'Save'}</button>
              </form>
            </div>
          </div>
        );
      }}
    </GameContext.Consumer>
  );
}
