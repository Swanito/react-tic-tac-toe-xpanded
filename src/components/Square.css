.square {
    width: 5em;
    height: 5em;
    border: 2px solid black;
    text-align: center;
    display: flex;
    align-items: center; 
    justify-content: center;
    background-color: #eee;
    margin: 0 1em 1em 1em
}

.square:hover {
    background-color: #ddd;
}