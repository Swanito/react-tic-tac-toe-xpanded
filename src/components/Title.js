import React from 'react';

const styles = {
    textAlign: 'center'
}

export default function Title() {
  return (
    <div style={styles}>
      <h1>Tic-Tac-Toe Xpanded!</h1>
    </div>
  );
}
