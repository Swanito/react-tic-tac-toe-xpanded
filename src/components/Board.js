import React, { useEffect, useState, useContext } from 'react';
import { GameContext } from '../context/GameContext';
import { calculateWinner as isPlayerWinner } from '../helpers/calculate-winner';
import Square from '../components/Square';
import './Square.css';

export default function Board() {
  const {
    boardSize,
    currentTurnPlayer,
    setCurrentTurnPlayer,
    winner,
    setWinner,
  } = useContext(GameContext);
  const prevPlayer = currentTurnPlayer === 'X' ? 'O' : 'X';

  const [owners, setOwners] = useState(
    Array.apply(null, Array(Number(boardSize * boardSize))),
  );

  const styles = {
    display: 'grid',
    gridTemplateColumns: `repeat(${boardSize}, minmax(6em, 6em))`,
    justifyContent: 'center',
    width: '20em',
    margin: '0 auto',
    pointerEvents: winner ? 'none' : 'auto',
  };

  const winnerStyle = {
      textAlign: 'center',
      margin: '1em'
  }

  function selectSquare(index) {
    if (owners[index]) return;
    const ownersCopy = [...owners];
    ownersCopy[index] = currentTurnPlayer;
    setOwners(ownersCopy);
    setCurrentTurnPlayer(currentTurnPlayer === 'X' ? 'O' : 'X');
  }

  useEffect(() => {
    if (isPlayerWinner(owners, boardSize, prevPlayer)) {
      setWinner(prevPlayer);
    }
  });

  return (
    <>
      <div style={winnerStyle}>{winner ? `THE WINNER IS ${winner}!` : ''}</div>
      <div style={styles}>
        {owners.map((_, i) => (
          <div className="square" onClick={() => selectSquare(i)} key={i}>
            {' '}
            {owners[i] === 'X' ? (
              <Square value={owners[i] ? owners[i] : ''} color="red" />
            ) : (
              <Square value={owners[i] ? owners[i] : ''} color="green" />
            )}{' '}
          </div>
        ))}
      </div>
    </>
  );
}
