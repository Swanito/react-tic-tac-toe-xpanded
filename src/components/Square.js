import React from 'react';
import './Square.css';

export default function Square({value, color}) {
    const styles = {
        color: color,
        fontWeight: 'bold',
        fontSize: '2em'
    }
    return (
        <div style={styles}>
            {value}
        </div>
    )
}
