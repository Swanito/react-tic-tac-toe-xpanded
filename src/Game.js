import Board from './components/Board';
import Settings from './components/Settings';
import Title from './components/Title';
import React, { useEffect, useState } from 'react';
import { GameContext } from './context/GameContext';

function Game() {
  const currentBoardSize = window.localStorage.getItem('boardSize') || 3;
  const currentPlayers = window.localStorage.getItem('players') || 2;

  const [currentTurnPlayer, setCurrentTurnPlayer] = useState('X');
  const [boardSize, setBoardSize] = useState(currentBoardSize);
  const [players, setPlayers] = useState(currentPlayers);
  const [winner, setWinner] = useState(null);

  useEffect(() => {
    window.localStorage.setItem('boardSize', boardSize);
    window.localStorage.setItem('players', players);
  }, [boardSize, players]);

  const settings = {
    boardSize,
    players,
    currentTurnPlayer,
    winner,
    setWinner,
    setCurrentTurnPlayer,
    setBoardSize,
    setPlayers,
  };

  return (
      <GameContext.Provider value={settings}>
        <Title />
        <Board />
        <Settings />
      </GameContext.Provider>
  );
}

export default Game;
