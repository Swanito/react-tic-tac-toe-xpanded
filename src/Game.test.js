import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Board from './components/Board';
import Title from './components/Title';
import Settings from './components/Settings';
import Square from './components/Square';
import Game from './Game';

test('renders Game Component', () => {
  const GameComponent = renderer.create(<Game />).toJSON();
  expect(GameComponent).toMatchSnapshot();
});

test('renders Title Component', () => {
  const GameComponent = renderer.create(<Title />).toJSON();
  expect(GameComponent).toMatchSnapshot();
});

test('renders Board Component', () => {
  const GameComponent = renderer.create(<Board />).toJSON();
  expect(GameComponent).toMatchSnapshot();
});

test('renders Settings Component', () => {
  const GameComponent = renderer.create(<Settings />).toJSON();
  expect(GameComponent).toMatchSnapshot();
});

test('renders Square Component', () => {
  const GameComponent = renderer.create(<Square />).toJSON();
  expect(GameComponent).toMatchSnapshot();
});