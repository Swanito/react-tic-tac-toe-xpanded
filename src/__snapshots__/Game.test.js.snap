// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Board Component 1`] = `
Array [
  <div
    style={
      Object {
        "margin": "1em",
        "textAlign": "center",
      }
    }
  >
    
  </div>,
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "repeat(1, minmax(6em, 6em))",
        "justifyContent": "center",
        "margin": "0 auto",
        "pointerEvents": "auto",
        "width": "20em",
      }
    }
  >
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
  </div>,
]
`;

exports[`renders Game Component 1`] = `
Array [
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <h1>
      Tic-Tac-Toe Xpanded!
    </h1>
  </div>,
  <div
    style={
      Object {
        "margin": "1em",
        "textAlign": "center",
      }
    }
  >
    
  </div>,
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "repeat(3, minmax(6em, 6em))",
        "justifyContent": "center",
        "margin": "0 auto",
        "pointerEvents": "auto",
        "width": "20em",
      }
    }
  >
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
    <div
      className="square"
      onClick={[Function]}
    >
       
      <div
        style={
          Object {
            "color": "green",
            "fontSize": "2em",
            "fontWeight": "bold",
          }
        }
      >
        
      </div>
       
    </div>
  </div>,
  <div
    style={
      Object {
        "alignItems": "flex-start",
        "display": "flex",
        "flexDirection": "column",
        "margin": "0 auto",
        "width": "20em",
      }
    }
  >
    <div>
      PLAYERS: 
      2
       - BOARD SIZE: 
      3
      *
      3
    </div>
    <br />
    <div>
      IT'S YOUR TURN 
      X
    </div>
    <br />
    <div>
      <form
        onSubmit={[Function]}
      >
        <label
          htmlFor="board-size-select"
        >
          Change board Size
        </label>
        <select
          onChange={[Function]}
          value={3}
        >
          <option
            value={3}
          >
            3*3
          </option>
          <option
            value={6}
          >
            6*6
          </option>
          <option
            value={9}
          >
            9*9
          </option>
        </select>
        <button
          type="submit"
        >
          Save
        </button>
      </form>
    </div>
  </div>,
]
`;

exports[`renders Settings Component 1`] = `
<div
  style={
    Object {
      "alignItems": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "margin": "0 auto",
      "width": "20em",
    }
  }
>
  <div>
    PLAYERS: 
    2
     - BOARD SIZE: 
    1
    *
    1
  </div>
  <br />
  <div>
    IT'S YOUR TURN 
  </div>
  <br />
  <div>
    <form
      onSubmit={[Function]}
    >
      <label
        htmlFor="board-size-select"
      >
        Change board Size
      </label>
      <select
        onChange={[Function]}
        value={1}
      >
        <option
          value={3}
        >
          3*3
        </option>
        <option
          value={6}
        >
          6*6
        </option>
        <option
          value={9}
        >
          9*9
        </option>
      </select>
      <button
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</div>
`;

exports[`renders Square Component 1`] = `
<div
  style={
    Object {
      "color": undefined,
      "fontSize": "2em",
      "fontWeight": "bold",
    }
  }
/>
`;

exports[`renders Title Component 1`] = `
<div
  style={
    Object {
      "textAlign": "center",
    }
  }
>
  <h1>
    Tic-Tac-Toe Xpanded!
  </h1>
</div>
`;
